rs/messaging/tests/memory_tests.rs

```diff
@@ -0,0 +1,500 @@
+/// 'chatter' enabled, i.e. the installed canisters aremaking random calls (including downstream calls).
+///
+/// For the second phase, the 'chatter' is disabled and additional rounds are executed until all
+/// hanging calls have received a reply.
+///
+/// Checks the guaranteed response message memory never exceeds the limit, all calls eventually
+/// receive a reply (or are rejected synchronously when attempting them), the message memory goes
+/// back to 0 after all in-flight messages have been dealt with.
+fn check_guaranteed_response_message_memory_limits_are_respected_impl(
+    seeds: &[u64],
+    max_payload_bytes: u32,
+    calls_per_round: u32,
+    reply_weight: u32,
+    call_weight: u32,
+) -> Result<(), (String, DebugInfo)> {
+    // The number of rounds to execute while chatter is on.
+    const CHATTER_PHASE_ROUND_COUNT: u64 = 30;
+    // The maximum number of rounds to execute after chatter is turned off. It it takes more than
+    // this number of rounds until there are no more hanging calls, the test fails.
+    const SHUTDOWN_PHASE_MAX_ROUNDS: u64 = 300;
+    // The amount of memory available for guaranteed response message memory on `local_env`.
+    const LOCAL_MESSAGE_MEMORY_CAPACITY: u64 = 100 * MB;
+    // The amount of memory available for guaranteed response message memory on `remote_env`.
+    const REMOTE_MESSAGE_MEMORY_CAPACITY: u64 = 50 * MB;
+
+    let fixture = Fixture::new(FixtureConfig {
+        local_canisters_count: 2,
+        local_max_instructions_per_round: 100_000_000,
+        local_message_memory_capacity: LOCAL_MESSAGE_MEMORY_CAPACITY,
+        remote_canisters_count: 1,
+        remote_max_instructions_per_round: 100_000_000,
+        remote_message_memory_capacity: REMOTE_MESSAGE_MEMORY_CAPACITY,
+    });
+
+    let config = CanisterConfig::try_new(
+        fixture.canisters(),   // receivers
+        0..=max_payload_bytes, // call_bytes
+        0..=max_payload_bytes, // reply_bytes
+        0..=0,                 // instructions_count
+    )
+    .unwrap();
+
+    // Send configs to canisters, seed the rng.
+    for (index, canister) in fixture.canisters().into_iter().enumerate() {
+        fixture.replace_config(canister, config.clone()).unwrap();
+        fixture.seed_rng(canister, seeds[index]);
+        fixture
+            .replace_reply_weight(canister, reply_weight)
+            .unwrap();
+        fixture.replace_call_weight(canister, call_weight).unwrap();
+    }
+
+    // Start chatter on all canisters.
+    fixture.start_chatter(calls_per_round).unwrap();
+
+    // Build up backlog and keep up chatter for while.
+    for _ in 0..CHATTER_PHASE_ROUND_COUNT {
+        fixture.tick();
+
+        // Check message memory limits are respected.
+        let (local_memory, remote_memory) = fixture.guaranteed_response_message_memory_taken();
+        if local_memory > LOCAL_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Chatter: local memory exceeds limit");
```

Comment:
Personal preference: `debug_info()` is a big vague. `fail_with_message()`? `fail_with_reason()`?

Reply:
Called it `failed_with_reason()` since `message` could be confused for an inter canister message and this is supposed to be used after a failure is detected by other means, not fail a test like an `assert` would (therefore `failed_with` not `fail_with`).